sensing/pointcloud_preprocessor/src/outlier_filter/ring_outlier_filter_nodelet.cpp

```diff
@@ -61,122 +67,286 @@ void RingOutlierFilterComponent::faster_filter(
+  const float object_length_threshold2 = object_length_threshold_ * object_length_threshold_;
+  auto isCluster = [=](const WalkInfo & walk_info) {
+    // A cluster is a walk which has many points or is long enough
+    if (walk_info.num_points > num_points_threshold_) return true;
+
+    // Using the law of cosines, the distance between 2 points can be written as:
+    //   |p2-p1|^2 = d1^2 + d2^2 - 2*d1*d2*cos(a)
+    // where d1 and d2 are the distance attribute of p1 and p2, and 'a' the azimuth diff between
+    // the 2 points.
+    const float dist2 =
+      walk_info.first_point_distance * walk_info.first_point_distance +
+      walk_info.last_point_distance * walk_info.last_point_distance -
+      2 * walk_info.first_point_distance * walk_info.last_point_distance *
+        std::cos((walk_info.last_point_azimuth - walk_info.first_point_azimuth) * (M_PI / 18000.0));
+    return dist2 > object_length_threshold2;
+  };
+
+  // check if 2 points belong to the same walk
+  auto isSameWalk =
+    [=](float curr_distance, float curr_azimuth, float prev_distance, float prev_azimuth) {
+      float azimuth_diff = curr_azimuth - prev_azimuth;
+      azimuth_diff = azimuth_diff < 0.f ? azimuth_diff + 36000.f : azimuth_diff;
+      return std::max(curr_distance, prev_distance) <
+               std::min(curr_distance, prev_distance) * distance_ratio_ &&
+             azimuth_diff < 100.f;
+    };
+
+  // tmp vectors to keep track of walk/ring state while processing points in order (cache efficient)
+  std::vector<WalkInfo> walks;             // all walks
+  std::vector<RingWalkInfo> rings;         // info for each ring
+  std::vector<size_t> points_walk_id;      // each point walk idx
+  std::vector<bool> walks_cluster_status;  // for each generated walk, stores whether it is a cluter
+
+  size_t latest_walk_id = -1UL;  // ID given to the last walk created
+
+  walks.reserve(max_rings_num_ * 2);  // only store first and last walk for each ring
+  // initialize each ring with empty walk
+  rings.resize(max_rings_num_, RingWalkInfo{nullptr, nullptr});
+  // points are initially associated to no walk
+  points_walk_id.resize(input->width * input->height, -1UL);
+  walks_cluster_status.reserve(
+    max_rings_num_ * 2);  // In the worst case, this could grow to the number of input points
+
+  int invalid_ring_count = 0;
+
+  // Build walks and classify points
+  for (const auto & [raw_p, point_walk_id] :
+       ranges::views::zip(input->data | ranges::views::chunk(input->point_step), points_walk_id)) {
+    const uint16_t ring_idx = getPointRing(raw_p.data());
+    const float curr_azimuth = getPointAzimuth(raw_p.data());
+    const float curr_distance = getPointDistance(raw_p.data());
+
+    if (ring_idx >= max_rings_num_) {
+      // Either the data is corrupted or max_rings_num_ is not set correctly
+      // Note: point_walk_id == -1 so the point will be filtered out
+      invalid_ring_count += 1;
+      continue;
+    }
```

Comment:
I have now changed the implementation so we don't have this unsafety anymore:
`WalkInfo` structs are now directly stored in `RingWalkInfo` (because we only need to store 2 walks per ring).
This also resulted in another significant performance gain (see comment below).